[Bug] After Upgrading Jetty12 - Doris FE won't start

### Search before asking

- [x] I had searched in the [issues](https://github.com/apache/doris/issues?q=is%3Aissue) and found no similar issues.


### Version

# Doris v3.1.4 FE fails to start when `enable_https = true`

## Summary

After upgrading to **Doris v3.1.4**, the Frontend (FE) fails to start when HTTPS is enabled:

```properties
enable_https = true
```

### Error logs: 

```2026-01-08 07:37:36,441 INFO (UNKNOWN fe_9a2e4182_39db_44b7_9446_d71915e1bbdd(-1)|1) [ServletContextHandler.doStop():1117] Stopped osbwej.JettyEmbeddedWebAppContext@bfd7235{application,/,b=file:/tmp/jetty-docbase.8050.1450579242224439361/,a=AVAILABLE,h=oeje10s.SessionHandler@70027aec{STOPPED}}
2026-01-08 07:37:36,466 WARN (UNKNOWN fe_9a2e4182_39db_44b7_9446_d71915e1bbdd(-1)|1) [AbstractApplicationContext.refresh():635] Exception encountered during context initialization - cancelling refresh attempt: org.springframework.context.ApplicationContextException: Failed to start bean 'webServerStartStop'
2026-01-08 07:37:36,476 INFO (UNKNOWN fe_9a2e4182_39db_44b7_9446_d71915e1bbdd(-1)|1) [ConditionEvaluationReportLogger.logMessage():82] 

Error starting ApplicationContext. To display the condition evaluation report re-run your application with 'debug' enabled.
2026-01-08 07:37:36,493 ERROR (UNKNOWN fe_9a2e4182_39db_44b7_9446_d71915e1bbdd(-1)|1) [SpringApplication.reportFailure():857] Application run failed
org.springframework.context.ApplicationContextException: Failed to start bean 'webServerStartStop'
	at org.springframework.context.support.DefaultLifecycleProcessor.doStart(DefaultLifecycleProcessor.java:408)
	at org.springframework.context.support.DefaultLifecycleProcessor.doStart(DefaultLifecycleProcessor.java:394)
	at org.springframework.context.support.DefaultLifecycleProcessor$LifecycleGroup.start(DefaultLifecycleProcessor.java:586)
	at java.base/java.lang.Iterable.forEach(Iterable.java:75)
	at org.springframework.context.support.DefaultLifecycleProcessor.startBeans(DefaultLifecycleProcessor.java:364)
	at org.springframework.context.support.DefaultLifecycleProcessor.onRefresh(DefaultLifecycleProcessor.java:310)
	at org.springframework.context.support.AbstractApplicationContext.finishRefresh(AbstractApplicationContext.java:1009)
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:630)
	at org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.refresh(ServletWebServerApplicationContext.java:146)
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:752)
	at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:439)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:318)
	at org.springframework.boot.builder.SpringApplicationBuilder.run(SpringApplicationBuilder.java:149)
	at org.apache.doris.httpv2.HttpServer.start(HttpServer.java:191)
	at org.apache.doris.DorisFE.start(DorisFE.java:235)
	at org.apache.doris.DorisFE.main(DorisFE.java:107)
Caused by: org.springframework.boot.web.server.WebServerException: Unable to start embedded Jetty server
	at org.springframework.boot.web.embedded.jetty.JettyWebServer.start(JettyWebServer.java:177)
	at org.springframework.boot.web.servlet.context.WebServerStartStopLifecycle.start(WebServerStartStopLifecycle.java:44)
	at org.springframework.context.support.DefaultLifecycleProcessor.doStart(DefaultLifecycleProcessor.java:405)
	... 15 more
Caused by: java.lang.IllegalStateException: WebSocketComponents has not been created
	at org.eclipse.jetty.websocket.core.server.WebSocketServerComponents.getWebSocketComponents(WebSocketServerComponents.java:173)
	at org.eclipse.jetty.websocket.core.server.WebSocketServerComponents.getWebSocketComponents(WebSocketServerComponents.java:161)
	at org.eclipse.jetty.websocket.core.server.WebSocketMappings.ensureMappings(WebSocketMappings.java:66)
	at org.eclipse.jetty.ee10.websocket.servlet.WebSocketUpgradeFilter.init(WebSocketUpgradeFilter.java:209)
	at org.eclipse.jetty.ee10.servlet.FilterHolder.initialize(FilterHolder.java:136)
	at org.eclipse.jetty.ee10.servlet.ServletHandler.lambda$initialize$2(ServletHandler.java:666)
	at org.eclipse.jetty.util.ExceptionUtil.call(ExceptionUtil.java:339)
	at org.eclipse.jetty.util.ExceptionUtil$MultiException.callAndCatch(ExceptionUtil.java:277)
	at org.eclipse.jetty.ee10.servlet.ServletHandler.lambda$initialize$3(ServletHandler.java:663)
	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1625)
	at java.base/java.util.stream.Streams$ConcatSpliterator.forEachRemaining(Streams.java:734)
	at java.base/java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:762)
	at org.eclipse.jetty.ee10.servlet.ServletHandler.initialize(ServletHandler.java:686)
	at org.springframework.boot.web.embedded.jetty.JettyEmbeddedWebAppContext$JettyEmbeddedServletHandler.deferredInitialize(JettyEmbeddedWebAppContext.java:57)
	at org.eclipse.jetty.server.handler.ContextHandler$ScopedContext.call(ContextHandler.java:1488)
	at org.springframework.boot.web.embedded.jetty.JettyEmbeddedWebAppContext.deferredInitialize(JettyEmbeddedWebAppContext.java:42)
	at org.springframework.boot.web.embedded.jetty.JettyWebServer.handleDeferredInitialize(JettyWebServer.java:240)
	at org.springframework.boot.web.embedded.jetty.JettyWebServer.handleDeferredInitialize(JettyWebServer.java:243)
	at org.springframework.boot.web.embedded.jetty.JettyWebServer.start(JettyWebServer.java:154)
	... 17 more[1:12 PM]2026-01-08 07:37:36,496 ERROR (UNKNOWN fe_9a2e4182_39db_44b7_9446_d71915e1bbdd(-1)|1) [Throwable$WrappedPrintStream.println():763] 	at org.springframework.boot.web.embedded.jetty.JettyEmbeddedWebAppContext.deferredInitialize(JettyEmbeddedWebAppContext.java:42)
2026-01-08 07:37:36,496 ERROR (UNKNOWN fe_9a2e4182_39db_44b7_9446_d71915e1bbdd(-1)|1) [Throwable$WrappedPrintStream.println():763] 	at org.springframework.boot.web.embedded.jetty.JettyWebServer.handleDeferredInitialize(JettyWebServer.java:240)
2026-01-08 07:37:36,496 ERROR (UNKNOWN fe_9a2e4182_39db_44b7_9446_d71915e1bbdd(-1)|1) [Throwable$WrappedPrintStream.println():763] 	at org.springframework.boot.web.embedded.jetty.JettyWebServer.handleDeferredInitialize(JettyWebServer.java:243)
2026-01-08 07:37:36,496 ERROR (UNKNOWN fe_9a2e4182_39db_44b7_9446_d71915e1bbdd(-1)|1) [Throwable$WrappedPrintStream.println():763] 	at org.springframework.boot.web.embedded.jetty.JettyWebServer.start(JettyWebServer.java:154)
2026-01-08 07:37:36,496 ERROR (UNKNOWN fe_9a2e4182_39db_44b7_9446_d71915e1bbdd(-1)|1) [Throwable$WrappedPrintStream.println():763] 	... 17 more
2026-01-08 07:37:36,496 ERROR (UNKNOWN fe_9a2e4182_39db_44b7_9446_d71915e1bbdd(-1)|1) [DorisFE.start():259] 
org.springframework.context.ApplicationContextException: Failed to start bean 'webServerStartStop'
	at org.springframework.context.support.DefaultLifecycleProcessor.doStart(DefaultLifecycleProcessor.java:408)
	at org.springframework.context.support.DefaultLifecycleProcessor.doStart(DefaultLifecycleProcessor.java:394)
	at org.springframework.context.support.DefaultLifecycleProcessor$LifecycleGroup.start(DefaultLifecycleProcessor.java:586)
	at java.base/java.lang.Iterable.forEach(Iterable.java:75)
	at org.springframework.context.support.DefaultLifecycleProcessor.startBeans(DefaultLifecycleProcessor.java:364)
	at org.springframework.context.support.DefaultLifecycleProcessor.onRefresh(DefaultLifecycleProcessor.java:310)
	at org.springframework.context.support.AbstractApplicationContext.finishRefresh(AbstractApplicationContext.java:1009)
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:630)
	at org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.refresh(ServletWebServerApplicationContext.java:146)
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:752)
	at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:439)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:318)
	at org.springframework.boot.builder.SpringApplicationBuilder.run(SpringApplicationBuilder.java:149)
	at org.apache.doris.httpv2.HttpServer.start(HttpServer.java:191)
	at org.apache.doris.DorisFE.start(DorisFE.java:235)
	at org.apache.doris.DorisFE.main(DorisFE.java:107)
Caused by: org.springframework.boot.web.server.WebServerException: Unable to start embedded Jetty server
	at org.springframework.boot.web.embedded.jetty.JettyWebServer.start(JettyWebServer.java:177)
	at org.springframework.boot.web.servlet.context.WebServerStartStopLifecycle.start(WebServerStartStopLifecycle.java:44)
	at org.springframework.context.support.DefaultLifecycleProcessor.doStart(DefaultLifecycleProcessor.java:405)
	... 15 more
Caused by: java.lang.IllegalStateException: WebSocketComponents has not been created
	at org.eclipse.jetty.websocket.core.server.WebSocketServerComponents.getWebSocketComponents(WebSocketServerComponents.java:173)
	at org.eclipse.jetty.websocket.core.server.WebSocketServerComponents.getWebSocketComponents(WebSocketServerComponents.java:161)
	at org.eclipse.jetty.websocket.core.server.WebSocketMappings.ensureMappings(WebSocketMappings.java:66)
	at org.eclipse.jetty.ee10.websocket.servlet.WebSocketUpgradeFilter.init(WebSocketUpgradeFilter.java:209)
	at org.eclipse.jetty.ee10.servlet.FilterHolder.initialize(FilterHolder.java:136)
	at org.eclipse.jetty.ee10.servlet.ServletHandler.lambda$initialize$2(ServletHandler.java:666)
	at org.eclipse.jetty.util.ExceptionUtil.call(ExceptionUtil.java:339)
	at org.eclipse.jetty.util.ExceptionUtil$MultiException.callAndCatch(ExceptionUtil.java:277)
	at org.eclipse.jetty.ee10.servlet.ServletHandler.lambda$initialize$3(ServletHandler.java:663)
	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1625)
	at java.base/java.util.stream.Streams$ConcatSpliterator.forEachRemaining(Streams.java:734)
	at java.base/java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:762)
	at org.eclipse.jetty.ee10.servlet.ServletHandler.initialize(ServletHandler.java:686)
	at org.springframework.boot.web.embedded.jetty.JettyEmbeddedWebAppContext$JettyEmbeddedServletHandler.deferredInitialize(JettyEmbeddedWebAppContext.java:57)
	at org.eclipse.jetty.server.handler.ContextHandler$ScopedContext.call(ContextHandler.java:1488)
	at org.springframework.boot.web.embedded.jetty.JettyEmbeddedWebAppContext.deferredInitialize(JettyEmbeddedWebAppContext.java:42)
	at org.springframework.boot.web.embedded.jetty.JettyWebServer.handleDeferredInitialize(JettyWebServer.java:240)
	at org.springframework.boot.web.embedded.jetty.JettyWebServer.handleDeferredInitialize(JettyWebServer.java:243)
	at org.springframework.boot.web.embedded.jetty.JettyWebServer.start(JettyWebServer.java:154)
	... 17 more
2026-01-08 07:37:36,496 INFO (Thread-0|27) [DorisFE.lambda$start$0():157] Received shutdown signal, starting graceful shutdown...
2026-01-08 07:37:36,496 INFO (Thread-0|27) [DorisFE.gracefulShutdown():585] graceful shutdown finished
```

### What's Wrong?

FE fails during startup with Jetty initialization errors related to WebSocket components and immediately triggers a graceful shutdown.

### What You Expected?

FE should start normally with HTTPS enabled.

### How to Reproduce?

Use v3.1.4 and enable_https=true and fe.conf and start

### Anything Else?

_No response_

### Are you willing to submit PR?

- [x] Yes I am willing to submit a PR!

### Code of Conduct

- [x] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

--- a/fe/fe-core/src/main/java/org/apache/doris/httpv2/config/WebServerFactoryCustomizerConfig.java
+++ b/fe/fe-core/src/main/java/org/apache/doris/httpv2/config/WebServerFactoryCustomizerConfig.java
@@ -19,6 +19,10 @@ package org.apache.doris.httpv2.config;
 
 import org.apache.doris.common.Config;
 
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.eclipse.jetty.ee10.webapp.WebAppContext;
+import org.eclipse.jetty.ee10.websocket.server.config.JettyWebSocketServletContainerInitializer;
 import org.eclipse.jetty.server.HttpConfiguration;
 import org.eclipse.jetty.server.HttpConnectionFactory;
 import org.eclipse.jetty.server.ServerConnector;
@@ -31,26 +35,40 @@ import java.util.Collections;
 
 @Configuration
 public class WebServerFactoryCustomizerConfig implements WebServerFactoryCustomizer<ConfigurableJettyWebServerFactory> {
+    private static final Logger LOG = LogManager.getLogger(WebServerFactoryCustomizerConfig.class);
+
     @Override
     public void customize(ConfigurableJettyWebServerFactory factory) {
+
+        ((JettyServletWebServerFactory) factory).addServerCustomizers(server -> {
+            WebAppContext context = server.getDescendant(WebAppContext.class);
+            if (context != null) {
+                try {
+                    JettyWebSocketServletContainerInitializer.configure(context, null);
+                } catch (Exception e) {
+                    LOG.error("Failed to initialize WebSocket support", e);
+                    throw new RuntimeException("Failed to initialize WebSocket support", e);
+                }
+            }
+        });
+
         if (Config.enable_https) {
             ((JettyServletWebServerFactory) factory).setConfigurations(
                     Collections.singletonList(new HttpToHttpsJettyConfig())
             );
 
-            factory.addServerCustomizers(
-                    server -> {
-                        HttpConfiguration httpConfiguration = new HttpConfiguration();
-                        httpConfiguration.setSecurePort(Config.https_port);
-                        httpConfiguration.setSecureScheme("https");
-
-                        ServerConnector connector = new ServerConnector(server);
-                        connector.addConnectionFactory(new HttpConnectionFactory(httpConfiguration));
-                        connector.setPort(Config.http_port);
-
-                        server.addConnector(connector);
+            factory.addServerCustomizers(server -> {
+                if (server.getConnectors() != null && server.getConnectors().length > 0) {
+                    ServerConnector existingConnector = (ServerConnector) server.getConnectors()[0];
+                    HttpConnectionFactory httpFactory =
+                            existingConnector.getConnectionFactory(HttpConnectionFactory.class);
+                    if (httpFactory != null) {
+                        HttpConfiguration httpConfig = httpFactory.getHttpConfiguration();
+                        httpConfig.setSecurePort(Config.https_port);
+                        httpConfig.setSecureScheme("https");
                     }
-            );
+                }
+            });
         }
     }
 }

